### PR TITLE
Battery Implementation

### PR DESCRIPTION
**Pull request by jamesmontemagno**
### Description of Change ###

This implements Battery #20 

### API Changes ###

Added: 
 
- double - double Battery.ChargeLevel { get; } 
- BatteryState- BatteryState Battery.State{ get; } 
- BatteryPowerSource- BatteryPowerSource Battery.PowerSource { get; } 

### Behavioral Changes ###

None

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Caboodle/Battery/Battery.android.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        }
+        public static double ChargeLevel
+        {
+            get
+            {
+                ValidateBatteryStatsPermission();
+
+                using (var filter = new IntentFilter(Intent.ActionBatteryChanged))
+                {
+                    using (var battery = Platform.CurrentContext.RegisterReceiver(null, filter))
```
Comment (mattleibow) [inline Caboodle/Battery/Battery.android.cs]: You don't need multiple indents:

```csharp
using (var filter = new IntentFilter(Intent.ActionBatteryChanged)) 
using (var battery = Platform.CurrentContext.RegisterReceiver(null, filter))
{
}
```

File: Caboodle/Battery/Battery.android.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        }
+        public static BatteryState State
+        {
+            get
+            {
+                ValidateBatteryStatsPermission();
+                using (var filter = new IntentFilter(Intent.ActionBatteryChanged))
+                {
+                    using (var battery = Platform.CurrentContext.RegisterReceiver(null, filter))
+                    {
```
Comment (mattleibow) [inline Caboodle/Battery/Battery.android.cs]: nested usings.

File: Caboodle/Battery/Battery.android.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        }
+        public static BatteryPowerSource PowerSource
+        {
+            get
+            {
+                ValidateBatteryStatsPermission();
+                using (var filter = new IntentFilter(Intent.ActionBatteryChanged))
+                {
+                    using (var battery = Platform.CurrentContext.RegisterReceiver(null, filter))
+                    {
```
Comment (mattleibow) [inline Caboodle/Battery/Battery.android.cs]: nested usings

File: Caboodle/Battery/Battery.ios.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                    switch (UIDevice.CurrentDevice.BatteryState)
+                    {
+                        case UIDeviceBatteryState.Charging:
+                            return BatteryState.Charging;
+                        case UIDeviceBatteryState.Full:
+                            return BatteryState.Full;
+                        case UIDeviceBatteryState.Unplugged:
+                            return BatteryState.Discharging;
+                        default:
+                            return BatteryState.Unknown;
```
Comment (mattleibow) [inline Caboodle/Battery/Battery.ios.cs]: We could have a check here to see if `ChargeLevel == 100`, then return full.

File: Caboodle/Battery/Battery.netstandard.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Microsoft.Caboodle
+{
+    public static partial class Battery
+    {
+        public static double ChargeLevel { get; } = -1;
+        public static BatteryState State { get; } = BatteryState.Unknown;
+        public static BatteryPowerSource PowerSource { get; } = BatteryPowerSource.Unknown;
+    }
```
Comment (mattleibow) [inline Caboodle/Battery/Battery.netstandard.cs]: Should we just always throw here?

File: Caboodle/Battery/Battery.android.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        {
+            if (hasBatteryStatsPermission)
+                return;
+
+            var permission = Manifest.Permission.BatteryStats;
+            if (!Platform.HasPermissionInManifest(permission))
+                throw new PermissionException(permission);
+
+            hasBatteryStatsPermission = true;
+        }
```
Comment (mattleibow) [inline Caboodle/Battery/Battery.android.cs]: What happens here if it is in the manifest, but the user has denied permission? Can the user deny permission?

File: Caboodle/Battery/Battery.shared.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        Full,
+        NotCharging
+    }
+
+    public enum BatteryPowerSource
+    {
+        Unknown,
+        Battery,
+        AC,
+        USB,
```
Comment (mattleibow) [inline Caboodle/Battery/Battery.shared.cs]: What are the rules for naming these? Always use Pascal case or do we allow acronyms all upper case? I know stylecop will complain.

File: Caboodle/Battery/Battery.uwp.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        public static BatteryState State
+        {
+            get
+            {
+                var percentage = ChargeLevel;
+
+                if (percentage >= 1.0)
+                    return BatteryState.Full;
+                if (percentage < 0)
+                    return BatteryState.Unknown;
```
Comment (mattleibow) [inline Caboodle/Battery/Battery.uwp.cs]: I think this should go after the state checks. Could it be the case that it is at 0, and charging.

File: Caboodle/Shared/Exceptions.shared.cs [hunk 9]
```diff
@@ -2,11 +2,19 @@
 
 namespace Microsoft.Caboodle
 {
-    public class NotImplentedInReferenceAssembly : NotImplementedException
+    public class NotImplentedInReferenceAssemblyException : NotImplementedException
```
Comment (mattleibow) [inline Caboodle/Shared/Exceptions.shared.cs]: Will it be possible to move the rename into a separate PR so as not to mix things? Not essential, but something nice.

File: Samples/Caboodle.Samples/View/BatteryPage.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+using Xamarin.Forms;
+using Xamarin.Forms.Xaml;
+
+namespace Caboodle.Samples.View
+{
+	[XamlCompilation(XamlCompilationOptions.Compile)]
+	public partial class BatteryPage : ContentPage
```
Comment (mattleibow) [inline Samples/Caboodle.Samples/View/BatteryPage.xaml.cs]: This attribute is already on the assembly in App.xml.cs

File: Caboodle/Battery/Battery.android.cs (hunk 1 above)
Comment (jamesmontemagno) [reply Caboodle/Battery/Battery.android.cs]: IDK that looks confusing to me. I like it nested just makes it cleaner to read in my honest opinion.

File: Caboodle/Battery/Battery.ios.cs (hunk 4 above)
Comment (jamesmontemagno) [reply Caboodle/Battery/Battery.ios.cs]: I think iOS returns Full if that is the case...

File: Caboodle/Battery/Battery.netstandard.cs (hunk 5 above)
Comment (jamesmontemagno) [reply Caboodle/Battery/Battery.netstandard.cs]: oh hmmm yeah I guess that is better :)

File: Caboodle/Battery/Battery.uwp.cs (hunk 8 above)
Comment (jamesmontemagno) [reply Caboodle/Battery/Battery.uwp.cs]: Sounds good

File: Caboodle/Shared/Exceptions.shared.cs (hunk 9 above)
Comment (jamesmontemagno) [reply Caboodle/Shared/Exceptions.shared.cs]: yup

File: Caboodle/Battery/Battery.android.cs (hunk 6 above)
Comment (jamesmontemagno) [reply Caboodle/Battery/Battery.android.cs]: This is not a runtime permission so it is okay.

File: DeviceTests/Caboodle.DeviceTests.Shared/Battery_Tests.cs [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        public void Charge_Level() =>
+            Assert.Equal(1.0, Battery.ChargeLevel);
+
+        [Fact]
+        public void Charge_State() =>
+            Assert.Equal(BatteryState.Full, Battery.State);
+
+        [Fact]
+        public void Charge_Power() =>
+            Assert.Equal(BatteryPowerSource.Ac, Battery.PowerSource);
```
Comment (mattleibow) [inline DeviceTests/Caboodle.DeviceTests.Shared/Battery_Tests.cs]: Will these tests pass on an disconnected device with a low battery? 😄

Comment (jamesmontemagno) [reply DeviceTests/Caboodle.DeviceTests.Shared/Battery_Tests.cs]: I don't think so :( My assumption here is we are testing on emulators? how  should we handle this? Anything but known?

Comment (mattleibow) [reply DeviceTests/Caboodle.DeviceTests.Shared/Battery_Tests.cs]: yeah... you could try checking if level > 0... It probably will always be on ac/usb for tests as it has to be connected to deploy/run the tests.

File: Caboodle/Platform/Platform.android.cs [hunk 12]
```diff
@@ -20,6 +23,23 @@ public static void Init(Activity activity, Bundle bundle)
             lifecycleListener = new ActivityLifecycleContextListener();
             activity.Application.RegisterActivityLifecycleCallbacks(lifecycleListener);
         }
+
```
Comment (Redth) [inline Caboodle/Platform/Platform.android.cs]: Are we ready to expose this publicly?

File: Caboodle/Shared/Exceptions.shared.cs [hunk 13]
```diff
@@ -9,4 +9,12 @@ public NotImplentedInReferenceAssembly()
         {
         }
     }
+
+    public class PermissionException : Exception
```
Comment (Redth) [inline Caboodle/Shared/Exceptions.shared.cs]: I think this should subclass UnauthorizedAccessException

File: Caboodle/Platform/PlatformExtensions.uwp.cs [hunk 14]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Windows.Foundation;
+
+namespace Microsoft.Caboodle
+{
+    public static partial class PlatformExtensions
```
Comment (Redth) [inline Caboodle/Platform/PlatformExtensions.uwp.cs]: Should these be internal for now?

Comment (jamesmontemagno) [reply Caboodle/Platform/PlatformExtensions.uwp.cs]: Agreed!

File: Caboodle/Shared/Exceptions.shared.cs (hunk 13 above)
Comment (jamesmontemagno) [reply Caboodle/Shared/Exceptions.shared.cs]: Love it!

File: Caboodle/Platform/Platform.android.cs (hunk 12 above)
Comment (jamesmontemagno) [reply Caboodle/Platform/Platform.android.cs]: YEAH!

File: Caboodle/Platform/PlatformExtensions.uwp.cs (hunk 14 above)
Comment (jamesmontemagno) [reply Caboodle/Platform/PlatformExtensions.uwp.cs]: Done

File: Caboodle/Platform/Platform.android.cs [hunk 15]
```diff
@@ -18,12 +18,15 @@ public static partial class Platform
         internal static Activity CurrentActivity =>
             lifecycleListener?.Activity;
 
-        public static void Init(Activity activity, Bundle bundle)
+        public static void Init(Application application)
```
Comment (Redth) [inline Caboodle/Platform/Platform.android.cs]: I feel like having multiple overloads here is confusing... I'd personally prefer the original signature with Activity and Bundle, JUST IN CASE we ever need them.  This is what Forms does as well...

Comment (jamesmontemagno) [reply Caboodle/Platform/Platform.android.cs]: put back :) left Application overload in :)

File: Caboodle/Battery/Battery.shared.cs [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Caboodle
+{
+    public static partial class Battery
+    {
+        static event BatteryChangedEventHandler BatteryChanagedInternal;
+
+        public delegate void BatteryChangedEventHandler(BatteryChangedEventArgs e);
```
Comment (mattleibow) [inline Caboodle/Battery/Battery.shared.cs]: Maybe this delegate shouldn't be nested in a class. It will be long to type:

    new Battery.BatteryChangedEventHandler()

**Comment by mattleibow** [comment]
This looks good to merge!

File: Caboodle/Platform/Platform.android.cs [hunk 17]
```diff
@@ -52,15 +52,8 @@ class ActivityLifecycleContextListener : Java.Lang.Object, Application.IActivity
...
-        {
-            get
-            {
-                if (currentActivity.TryGetTarget(out var a))
-                    return a;
-                return null;
-            }
-        }
+        public Activity Activity =>
+            currentActivity.TryGetTarget(out var a) ? a : null;
```
Comment (mattleibow) [inline Caboodle/Platform/Platform.android.cs]: :heart:
